File: bd_chat_sdk/ios/Classes/BdChatSdkPlugin.swift
import Flutter
import UIKit
import BoldDeskChatSDK

public class BdChatSdkPlugin: NSObject, FlutterPlugin {
  private var channel: FlutterMethodChannel?
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "bd_chat_sdk", binaryMessenger: registrar.messenger())
    let instance = BdChatSdkPlugin()
    instance.channel = channel
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  private let platform: String = "Flutter"
  private let sdkVersion: String = "5.0.2"

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    
    switch call.method {
    case "configure":
      guard let args = call.arguments as? [String: Any],
        let appKey = args["appKey"] as? String,
        let brandUrl = args["brandUrl"] as? String,
        let culture = args["culture"] as? String
      else {
        result(["success": false, "message": "Missing arguments"])
        return
      }
      
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.configure(appKey: appKey, brandUrl: brandUrl, culture)
      result(nil)
    case "setPreferredTheme":
      guard let args = call.arguments as? [String: Any],
        let themeString = args["theme"] as? String
      else {
        result(["success": false, "message": "Missing theme"])
        return
      }

      let preferredTheme: SDKTheme
      switch themeString.lowercased() {
      case "light":
        preferredTheme = .light
      case "dark":
        preferredTheme = .dark
      default:
        preferredTheme = .system
      }

      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.setPreferredTheme(preferredTheme)
      result(nil)
    case "enablePushNotification":
      guard let args = call.arguments as? [String: Any],
        let token = args["fcmToken"] as? String
      else {
        return
      }
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.enablePushNotification(fcmToken: token ?? "")
      result(nil)   
    case "enableLogging":
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.enableLogging()
      result(nil) 
    case "isChatOpen":
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      let isChatOpen = BDChatSDK.isChatOpen()
      result(isChatOpen) 
    case "setPrefillFields":
      let args = call.arguments as? [String: Any]
      let name = args?["name"] as? String
      let email = args?["email"] as? String
      let phoneNumber = args?["phoneNumber"] as? String
      let rawChatFields = args?["chatFields"] as? [String: Any]
      var chatFields: [String: Any]? = nil
      if let raw = rawChatFields {
        var cleaned = [String: Any]()
        for (k, v) in raw {
          if !(v is NSNull) {
            cleaned[k] = v
          }
        }
        if !cleaned.isEmpty { chatFields = cleaned }
      }

      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.setPrefillFields(email: email, name: name, phoneNo: phoneNumber, fields: chatFields)
      result(nil)
    case "setUserToken":
      guard let args = call.arguments as? [String: Any],
        let userToken = args["userToken"] as? String
      else {
        return
      }
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.setUserToken(userToken)
      result(nil) 
    case "disablePushNotification":
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.disablePushNotification()
      result(nil) 
    case "showChat":
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.showChat()
      result(nil) 
    case "closeChat":
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.closeChat()
      result(nil) 
    case "clearSession":
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.clearSession()
      result(nil) 
    case "isFromChatSDK":
     if let args = call.arguments as? [String: Any],
        let messageData = args["messageData"] as? [AnyHashable: Any] {
          BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
          let isFromSDK = BDChatSDK.isFromChatSDK(userInfo: messageData)
        result(isFromSDK)
      } else {
          result(false)
      }
    case "applyCustomFontFamilyInIOS":
      guard let args = call.arguments as? [String: Any],
        let fontFamily = args["fontFamily"] as? String
      else {
        return
      }
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.customFontName = fontFamily
      result(nil)
    case "handlePushNotification":
      if let args = call.arguments as? [String: Any],
        let messageData = args["messageData"] as? [String: String] {

          BDChatSDK.handlePushNotification(userInfo: messageData)
          result(true)
      } else {
          result(false)
      }
    case "setSystemFontSize":
      guard let args = call.arguments as? [String: Any],
        let enable = args["enable"] as? Bool
      else {
        return
      }
      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.applySystemFontSize = enable
      result(nil)
    case "applyTheme":
      let args = call.arguments as? [String: Any]

      let appbarColor = args?["appbarColor"] as? String
      let accentColor = args?["accentColor"] as? String
      let backgroundColor = args?["backgroundColor"] as? String
      let stickyButtonColor = args?["stickyButtonColor"] as? String

      BDChatSDK.setPlatform(sdkPlatform: platform, sdkVersion: sdkVersion)
      BDChatSDK.applyTheme(
          appbarColor: appbarColor,
          accentColor: accentColor,
          backgroundColor: backgroundColor,
          stickyButtonColor: stickyButtonColor
      )
      result(nil)
    case "setOnTicketCreatedListener":
      let args = call.arguments as? [String: Any]
      let enabled = args?["enabled"] as? Bool ?? false

      if enabled {
          BDChatSDK.onTicketCreatedEventCallBack = { [weak self] ticketId in
              self?.channel?.invokeMethod(
                  "onTicketCreated",
                  arguments: ticketId
              )
          }
      } else {
          BDChatSDK.onTicketCreatedEventCallBack = nil
      }
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: bd_chat_sdk/android/src/main/kotlin/com/syncfusion/bolddeskchatsdk/bd_chat_sdk/BdChatSdkPlugin.kt
package com.syncfusion.bolddeskchatsdk.bd_chat_sdk

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.syncfusion.bolddeskandroidchatSDK.BoldDeskChatSDK
import com.syncfusion.bolddeskandroidchatSDK.R
import androidx.core.graphics.drawable.IconCompat
import android.graphics.Bitmap
import android.graphics.BitmapFactory

/** BdChatSdkPlugin */
class BdChatSdkPlugin : FlutterPlugin, MethodCallHandler {
    // The MethodChannel that will the communication between Flutter and native Android
    //
    // This local reference serves to register the plugin with the Flutter Engine and unregister it
    // when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "bd_chat_sdk")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall,result: Result ) {
        if (call.method == "configure") {
            val appKey = call.argument<String>("appKey") ?: ""
            val brandUrl = call.argument<String>("brandUrl") ?: ""
            val culture = call.argument<String>("culture") ?: "en-US"
            try {
                BoldDeskChatSDK.configure(context, appKey, brandUrl, culture)
            } catch (e: Exception) {
                result.error("INITIALIZATION_FAILED", e.message, null)
            }
        } else if (call.method == "showChat") {
            try {
                BoldDeskChatSDK.showChat(context)
                result.success("Chat Shown Successfully")
            } catch (e: Exception) {
                result.error("SHOW_CHAT_FAILED", e.message, null)
            }
        } else if (call.method == "closeChat") {
            try {
                BoldDeskChatSDK.closeChat(context)
                result.success("Chat Close Successfully")
            } catch (e: Exception) {
                result.error("CLOS_ECHAT_FAILED", e.message, null)
            }
        } else if (call.method == "clearSession") {
            try {
                BoldDeskChatSDK.clearSession()
                result.success("Clear Session Successfully")
            } catch (e: Exception) {
                result.error("CLEAR_SESSION_FAILED", e.message, null)
            }
        } else if (call.method == "enableLogging") {
            try {
                BoldDeskChatSDK.enableLogging()
                result.success("Enable Logging Successfully")
            } catch (e: Exception) {
                result.error("Enable_Logging_FAILED", e.message, null)
            }
        } else if (call.method == "setPrefillFields") {
            val name = call.argument<String>("name") ?: ""
            val email = call.argument<String>("email") ?: ""
            val phoneNumber = call.argument<String>("phoneNumber") ?: ""
            val chatFields = call.argument<Map<String, Any>>("chatFields")
            
            // Normalize: Convert Double to Int for whole numbers (both in arrays and standalone values)
            val normalizedChatFields = chatFields?.mapValues { entry ->
                when (val value = entry.value) {
                    is List<*> -> {
                        value.map { item ->
                            when (item) {
                                is Double -> if (item == item.toLong().toDouble()) item.toInt() else item
                                is Float -> if (item == item.toLong().toFloat()) item.toInt() else item
                                is Number -> {
                                    val n = item.toDouble()
                                    if (n == n.toLong().toDouble()) n.toInt() else n
                                }
                                else -> item
                            }
                        }
                    }
                    is Double -> if (value == value.toLong().toDouble()) value.toInt() else value
                    is Float -> if (value == value.toLong().toFloat()) value.toInt() else value
                    else -> value
                }
            }
            
            try {
                BoldDeskChatSDK.setPrefillFields(name, email, phoneNumber, normalizedChatFields)
                result.success("Prefill Fields Set Successfully")
            } catch (e: Exception) {
                result.error("PREFILL_FIELDS_FAILED", e.message, null)
            }
        } else if (call.method == "setUserToken") {
            val userToken = call.argument<String>("userToken") ?: ""
            try {
                BoldDeskChatSDK.setUserToken(userToken)
                result.success("UserToken Set Successfully")
            } catch (e: Exception) {
                result.error("USERTOKEN_SET_FAILED", e.message, null)
            }
        } else if (call.method == "setPreferredTheme") {
            val themeString = call.argument<String>("theme")?.lowercase() ?: ""
            val preferredTheme = when (themeString) {
                "light" -> BoldDeskChatSDK.SDKTheme.LIGHT
                "dark" -> BoldDeskChatSDK.SDKTheme.DARK
                else -> BoldDeskChatSDK.SDKTheme.SYSTEM
            }
            try {
                BoldDeskChatSDK.setPreferredTheme(preferredTheme)
                result.success("Theme Set Successfully")
            } catch (e: Exception) {
                result.error("THEME_SET_FAILED", e.message, null)
            }
        } else if (call.method == "applyCustomFontFamilyInAndroid") {
            val regular = call.argument<String>("regular") ?: ""
            val medium = call.argument<String>("medium") ?: ""
            val semiBold = call.argument<String>("semiBold") ?: ""
            val bold = call.argument<String>("bold") ?: ""
            try {
                // Dynamically find the resource ID for each font name
                val regularId = getFontResourceId(context, regular)
                val mediumId = getFontResourceId(context, medium)
                val semiBoldId = getFontResourceId(context, semiBold)
                val boldId = getFontResourceId(context, bold)
                BoldDeskChatSDK.applyCustomFontFamily(
                    regular = regularId,
                    medium = mediumId,
                    semiBold = semiBoldId,
                    bold = boldId
                )
                result.success(null)
            } catch (e: Exception) {
                result.error("FONT_NOT_FOUND", e.message, null)
            }
        } else if (call.method == "isFromChatSDK") {
            val messageData = call.argument<Map<String, String>>("messageData") ?: emptyMap()
            val isFromChatSDK = BoldDeskChatSDK.isFromChatSDK(messageData)
            result.success(isFromChatSDK)
        } else if (call.method == "enablePushNotification") {
            val fcmToken = call.argument<String>("fcmToken") ?: ""
            BoldDeskChatSDK.enablePushNotification(fcmToken)
            result.success(null)
        } else if (call.method == "disablePushNotification") {
            BoldDeskChatSDK.disablePushNotification()
            result.success(null)
        } else if (call.method == "handlePushNotifications") {
            val messageData = call.argument<Map<String, String>>("body") ?: emptyMap()
            val iconBytes = call.argument<ByteArray>("notificationIcon") ?: "".toByteArray()
            val bitmap = BitmapFactory.decodeByteArray(iconBytes, 0, iconBytes.size)
            val iconCompact = IconCompat.createWithBitmap(bitmap)
            try {
                BoldDeskChatSDK.handlePushNotifications(context, messageData, iconCompact)
                result.success("Notification shown Successfully")
            } catch (e: Exception) {
                result.error("NOTIFICATION_FAILED", e.message, null)
            }
        } else if (call.method == "applyTheme") {
            val appBarColor  = call.argument<String>("appbarColor") ?: ""
            val accentColor = call.argument<String>("accentColor") ?: ""
            val backgroundColor = call.argument<String>("backgroundColor") ?: ""
            val stickyButtonColor = call.argument<String>("stickyButtonColor") ?: ""
            BoldDeskChatSDK.applyTheme(context, appBarColor , accentColor, backgroundColor, stickyButtonColor)
            result.success("Theme applied Successfully")
        } else if (call.method == "setSystemFontSize"){
            val isEnabled = call.argument<Boolean>("enable") ?: false
            BoldDeskChatSDK.setSystemFontSize(isEnabled)
            result.success("Logging set Successfully")
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun getFontResourceId(context: Context, fontName: String): Int {
        // This is the magic: It looks for a resource of type "font" with the given name
        // in the application's package.
        val resourceId = context.resources.getIdentifier(fontName, "font", context.packageName)

        if (resourceId == 0) {
            // This error is critical for debugging!
            throw Exception("Font resource '$fontName' not found. " +
                    "Ensure it's in pubspec.yaml and the filename is correct.")
        }
        return resourceId
    }

    private fun resolveIcon(context: Context, iconName: String): Int {
        // Try to resolve from drawable resources
        val resId = context.resources.getIdentifier(
            iconName,
            "drawable",
            context.packageName
        )

        // Return resolved ID or default launcher icon
        return if (resId != 0) resId else android.R.drawable.ic_dialog_info
    }
}
